### PR TITLE
[JAX][DeepSeek]Implement SparseMatmul for DeepSeek 

### PR DESCRIPTION
**Pull request by bzgoogle**
# Description

Implement SparseMatmul for DeepSeek 

# Tests

The accuracy on MMLU for 10K prompts is [0.8619](https://paste.googleplex.com/5935578197655552),  which is aligned with densematmul(0.862)

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tests/models/jax/common/moe/test_deepseek_moe.py [hunk 1]
```diff
@@ -1,11 +1,16 @@
+import os
 import unittest
 
+os.environ['XLA_FLAGS'] = '--xla_force_host_platform_device_count=8'
```
Comment (jrplatin) [inline tests/models/jax/common/moe/test_deepseek_moe.py]: This won't have any adverse effects, right? Like affecting other tests?

File: tpu_commons/models/jax/common/moe/deepseek_moe.py [hunk 2]
```diff
@@ -109,3 +116,477 @@ def __post_init__(self, rngs: nnx.Rngs):
...
+        tile_size: A tuple (batch, activation_dim, weight_dim) for GMM tiling.
+        use_megablox: If True, uses the MegaBlox GMM kernel.
+        mesh: The device mesh.
+        # TODO: need to redesign this I/O for parallelism
+        num_expert_parallelism: The size of the 'expert' mesh dimension.
+        # TODO: determine if we get it from external or extrat it in MoE class
+        is_batch_sharded_by_expert: True if batch is sharded over 'expert' dim.
+    """
+    num_experts_per_tok: int
+    #TODO: tile size is (tile_batch_seq, tile_activation_dim, tile_weight_dim,) from MaxText
```
Comment (jrplatin) [inline tpu_commons/models/jax/common/moe/deepseek_moe.py:140]: I assume we'll address this later?

File: tpu_commons/models/jax/common/moe/deepseek_moe.py [hunk 3]
```diff
@@ -109,3 +116,477 @@ def __post_init__(self, rngs: nnx.Rngs):
...
+    def _unpermute(self, processed_tokens: jax.Array, sort_indices: jax.Array,
+                   router_weights_TX: jax.Array):
+        """Unsorts tokens to their original order and combines expert outputs with router's weight."""
+        with jax.named_scope("unpermute"):
+            unsorted_tokens_tD = self._sort_activations(
+                processed_tokens, jnp.argsort(sort_indices))
+            reshaped_tokens_TXD = unsorted_tokens_tD.reshape(
+                -1, self.num_experts_per_tok, self.hidden_size)
+        # jax.debug.print(
+        #     "✅ reshaped_tokens_TXD on device:  reshaped_tokens_TXD[5]={t}",
```
Comment (jrplatin) [inline tpu_commons/models/jax/common/moe/deepseek_moe.py]: Remove?

File: tpu_commons/models/jax/common/moe/deepseek_moe.py [hunk 4]
```diff
@@ -109,3 +116,477 @@ def __post_init__(self, rngs: nnx.Rngs):
...
+
+        else:
+            # --- NO EXPERT PARALLELISM ---
+            compute_inputs = sorted_inputs
+            compute_group_sizes = global_group_sizes
+            compute_expert_ids = global_sorted_experts
+            local_sorted_indices = jnp.arange(sorted_inputs.shape[0])
+
+        #debug_position_in_sorted = jnp.argsort(global_sort_indices)[40:48]
+        #debug_position_compute_inputs = jnp.argsort(local_sorted_indices)[debug_position_in_sorted]
```
Comment (jrplatin) [inline tpu_commons/models/jax/common/moe/deepseek_moe.py]: Remove?

File: tpu_commons/models/jax/deepseek_v3.py [hunk 5]
```diff
@@ -18,7 +18,8 @@
 from tpu_commons.models.jax.common.constants import KVCacheType
 from tpu_commons.models.jax.common.layers import (DenseFFW, Embedder, LMhead,
                                                   RMSNorm)
-from tpu_commons.models.jax.common.moe.deepseek_moe import DeepSeekV3Router
+from tpu_commons.models.jax.common.moe.deepseek_moe import (DeepSeekV3Router,
+                                                            SparseMoE)
```
Comment (jrplatin) [inline tpu_commons/models/jax/deepseek_v3.py:23]: I assume this implementation of Sparse Matmul is specific to DeepSeek right and shouldn't go in the `common` MoE class?

File: tpu_commons/models/jax/deepseek_v3.py [hunk 6]
```diff
@@ -79,6 +80,18 @@ def __init__(self,
 
         self.random_init = force_random_weights or self.vllm_config.additional_config.get(
             "random_weights", False)
+        self.sparse_matmul = self.vllm_config.additional_config.get(
```
Comment (jrplatin) [inline tpu_commons/models/jax/deepseek_v3.py:85]: Should it not be used by default?

File: tpu_commons/models/jax/deepseek_v3.py [hunk 7]
```diff
@@ -176,29 +189,56 @@ def _create_mla() -> MLA:
...
-                                    hidden_act=hidden_act,
-                                    hidden_size=hidden_size,
-                                    intermediate_size=ffw_intermediate_size,
-                                    rngs=self.rng,
-                                    random_init=self.random_init,
-                                    df_sharding=(None, ('model', 'expert')),
-                                    fd_sharding=(('model', 'expert'), None))
+                ed_sharding=('model', None),
+                e_sharding=('model', ))
+            if self.sparse_matmul:
```
Comment (jrplatin) [inline tpu_commons/models/jax/deepseek_v3.py:198]: Seems like `SparseMoE` and `MoE` have almost identical signatures, maybe just combine them by doing something like:

```
moe_cls = SparseMoE if self.sparse_matmul else MoE
moe_cls(...)
```
?

File: tests/models/jax/common/moe/test_deepseek_moe.py (hunk 1 above)
Comment (bzgoogle) [reply tests/models/jax/common/moe/test_deepseek_moe.py]: I think so. But to be safe, i will remove it

File: tpu_commons/models/jax/common/moe/deepseek_moe.py (hunk 2 above)
Comment (bzgoogle) [reply tpu_commons/models/jax/common/moe/deepseek_moe.py:140]: Correct, we will tune and optimize it later

File: tpu_commons/models/jax/deepseek_v3.py (hunk 5 above)
Comment (bzgoogle) [reply tpu_commons/models/jax/deepseek_v3.py:23]: similarly as the deepseek_v3_attent.py in common/attention, I currently put it under common/moe. Maybe we could refactor the structure in later PR to put model specific blocks in a split path

File: tests/models/jax/common/moe/test_deepseek_moe.py (hunk 1 above)
Comment (jrplatin) [reply tests/models/jax/common/moe/test_deepseek_moe.py]: Should be safe, so feel free to ignore

File: tpu_commons/models/jax/common/moe/deepseek_moe.py (hunk 3 above)
Comment (bzgoogle) [reply tpu_commons/models/jax/common/moe/deepseek_moe.py]: Done

File: tpu_commons/models/jax/common/moe/deepseek_moe.py (hunk 4 above)
Comment (bzgoogle) [reply tpu_commons/models/jax/common/moe/deepseek_moe.py]: Done

File: tpu_commons/models/jax/deepseek_v3.py (hunk 6 above)
Comment (bzgoogle) [reply tpu_commons/models/jax/deepseek_v3.py:85]: I would currently use the dense matmul as default, until we have complete the quantizaiton/sharding in your later PR

File: tpu_commons/models/jax/deepseek_v3.py (hunk 7 above)
Comment (bzgoogle) [reply tpu_commons/models/jax/deepseek_v3.py:198]: good idea. Currently the interface to SparseMoE is slightly different than MoE, which may need more efforts to combine them. But I put a TODO and will address it later.

Comment (jrplatin) [reply tpu_commons/models/jax/deepseek_v3.py:198]: Sounds good

File: tpu_commons/models/jax/deepseek_v3.py (hunk 5 above)
Comment (jrplatin) [reply tpu_commons/models/jax/deepseek_v3.py:23]: Sounds good, maybe make a TODO?
